init

werto87 opened the pull request:
emsdk_installer/2.0.10

This recipe creates a package which may be used as build_requires to build packages for emscripten wasm.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Croydon commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+class EmSDKInstallerConan(ConanFile):
+    name = "emsdk_installer"
+    description = "Emscripten is an Open Source LLVM to JavaScript compiler"
+    url = "https://github.com/bincrafters/conan-emsdk_installer"
```
I guess that can be counted as credits 😉

Croydon commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class EmSDKInstallerConan(ConanFile):
+    name = "emsdk_installer"
```
We don't append such packages with `_installer` anymore.
```suggestion
    name = "emsdk"
```

werto87 replied:
Is it acceptable that this package downloads its own node.js? I think I can make this package pass the conan hooks and still work. But is this wanted?

madebr replied:
Shouldn't it be `emscripten`?

Croydon reviewed (commented):
There was a previous attempt here https://github.com/conan-io/conan-center-index/pull/4860

werto87 commented:
Is it acceptable that this package downloads its own node.js? I think I can make this package pass the conan hooks and still work. But is this wanted?

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+    @staticmethod
+    def _touch(filename):
+        if not os.path.isfile(filename):
+            with open(filename, "w") as f:
+                f.write("\n")
```
```suggestion
            tools.save(filename, "\n")
```

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+
+    @staticmethod
+    def _create_dummy_file(directory):
+        if not os.path.isdir(directory):
+            os.makedirs(directory)
+        with open(os.path.join(directory, "dummy"), "w") as f:
+            f.write("\n")
```
I think `tools.save` automatically creates the directory structure.
As this method can be replaced with one function call, this method can be removed all together.
```suggestion
        tools.save(os.path.join(directory, "dummy"), "\n")
```

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+                              "set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)")
+
+    def _define_tool_var(self, name, value):
+        suffix = '.bat' if os.name == 'nt' else ''
```
Don't use `os`. `settings.os` determines the os on which `emscripten` is supposed to be built for.
```suggestion
        suffix = '.bat' if self.settings.os == "Windows"
```

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+    url = "https://github.com/bincrafters/conan-emsdk_installer"
+    homepage = "https://github.com/kripken/emscripten"
+    topics = ("conan", "emsdk", "emscripten", "installer", "sdk")
+    license = "MIT"
```
It needs access to the `os` on which it runs.
```suggestion
    license = "MIT"
    settings = "os"
```

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+    license = "MIT"
+
+    short_paths = True
+    _source_subfolder = "source_subfolder"
+
+    def source(self):
```
Remove `settings.os` in `package_info` to make sure that only one package is built with cross-os support.
```suggestion
    @property
    def _source_subfolder(self):
        return "source_subfolder"

    def package_id(self):
        del self.info.settings.os

    def source(self):
```

madebr commented on recipes/emsdk_installer/all/conanfile.py:
```diff
@@ -0,0 +1,121 @@
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_folder = "emsdk-%s" % self.version
+        os.rename(extracted_folder, self._source_subfolder)
+
```
A `validate` method to check whether `os` is valid would be useful, I think.
This recipe is supposed to be used in a cross build configuration. So it should also probably test whether the target os is actually `wasm`.
Something like:
```python
    def validate(self):
        if self.settings.os not in ("Linux", "Macos", "Windows"):
             raise ConanInvalidConfiguration("Invalid OS. Only Linux, Macos and Windows are supported.")
        if hasattr(self, "settings_target"):
            if self.settings_target.os != "wasm":
                 raise ConanInvalidConfiguration("Emscripten targets only wasm, not {}.".format(self.settings_target.os)
```
You will also need to add `from conans import ConanInvalidConfiguration` at the top of this file.

werto87 replied:
When providing "--profile:host" self.settings.os contains the value of that profile. So when the consumer sets a profile as host which os is Emscripten this does not work.

I think  "wasm" is not an os.

To be honest I do not know what configurations are wrong.
I think it is a great idea to give the user feedback.
Maybe we could use "self.settings_build" instead of "self.settings.os" to check if the build os is linux mac or windows.

madebr commented on recipes/emsdk_installer/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+conan_basic_setup()
+
+
+file(GLOB SOURCE_FILES *.cpp)
+
+add_executable(${PROJECT_NAME} ${SOURCE_FILES})
```
```suggestion
add_executable(${PROJECT_NAME} test_package.cpp)
```

madebr commented on recipes/emsdk_installer/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,18 @@
+        cmake.definitions["CONAN_DISABLE_CHECK_COMPILER"] = True
+        cmake.configure()
+        cmake.build()
+        #self.run("cmake --build %s --config Release" % self.build_folder)
```
```suggestion
```

madebr commented on recipes/emsdk_installer/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,18 @@
+
+    def test(self):
+        test_file = os.path.join(self.build_folder, "bin", "test_package.js")
+        self.run('node %s' % test_file, run_environment=True)
```
Conan-center-index has a `nodejs` recipe.
Perhaps you can add a build requirement on `nodejs` (only needed in the test package)?

werto87 replied:
There is a problem. The emsdk downloads its own nodejs and wants to use it. The recipe in bincrafters used a trick to supress this. But in the current versions I think since 2.0 its kinda hard. And it feels like fighting with emsdk :smiley: .
I want to provide a solution which allows people to use emsdk to compile other packages to emscripten wasm. Even if it is not so good to have dependencies not managed by conan. I want to let the dependencies problem open to a fix by somebody else with more motivation in this part. I am open to fix the your other remarks.

madebr replied:
Let's first fix the errors in https://github.com/conan-io/conan-center-index/pull/6163#issuecomment-873920734 so c3i can at least start building something.

werto87 commented:
The build fails with "CMakeFiles/test_package.dir/test_package.cpp.o: must specify -mwasm64 to process wasm64 object files"
I get the same error on my system if I do not set "--profile:build=clang --profile:host=emscripten". How to fix this on conan CI?
For reference:
cat .conan/profiles/emscripten
[settings]
os=Linux
arch=x86_64
compiler=clang
compiler.version=13
compiler.libcxx=libc++
[env]
CC=/usr/bin/clang
CXX=/usr/bin/clang++

cat .conan/profiles/clang
[settings]
os=Linux
arch=x86_64
compiler=clang
compiler.version=13
compiler.libcxx=libc++
[env]
CC=/usr/bin/clang
CXX=/usr/bin/clang++

madebr commented:
The `test_package` log contains:
```
cache:INFO: generating system library: libc++-noexcept.a... (this will be cached in "/home/conan/w/BuildSingleReference/.conan/data/emsdk_installer/2.0.10/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/.emscripten_cache/wasm/libc++-noexcept.a" for subsequent builds)
```
Is it possible to configure this cache folder? conan generates a manifest file after packaging, but it's always nice to keep the package folder pristine.

werto87 commented:
> The `test_package` log contains:
> 
> ```
> cache:INFO: generating system library: libc++-noexcept.a... (this will be cached in "/home/conan/w/BuildSingleReference/.conan/data/emsdk_installer/2.0.10/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/.emscripten_cache/wasm/libc++-noexcept.a" for subsequent builds)
> ```
> 
> Is it possible to configure this cache folder? conan generates a manifest file after packaging, but it's always nice to keep the package folder pristine.

It is possible to set the folder. Which path do you suggest and what is wrong with the current path?

madebr commented:
> 
> 
> The build fails with "CMakeFiles/test_package.dir/test_package.cpp.o: must specify -mwasm64 to process wasm64 object files"
> I get the same error on my system if I do not set "--profile:build=clang --profile:host=emscripten". How to fix this on conan CI?

This recipe is supposed to be used as a build requirement, so the test should be run with `emscripten` as a build requirement.
See https://github.com/conan-io/conan/pull/8787 on how to do this (new conan 1.36.0 feature)

I would try to write the test package such that, when not using conan's cross building feature, try to work around the missing bits.
How come cmake only adds `-mwasm64` to the link command when cross building? Is it possible to force it when not using conan's cross building support?


> It is possible to set the folder. Which path do you suggest and what is wrong with the current path?

I am asking because it caches in the package folder. 
To give a counter example: I am recently toying a bit with writing Android apps. Its build system, gradle, caches libraries in `~/.gradle`.

werto87 commented:
> 
> This recipe is supposed to be used as a build requirement, so the test should be run with `emscripten` as a build requirement.
> See [conan-io/conan#8787](https://github.com/conan-io/conan/pull/8787) on how to do this (new conan 1.36.0 feature)
> 
I added 'test_type = "build_requires"' to the test_package recipe.

> I would try to write the test package such that, when not using conan's cross building feature, try to work around the missing bits.
> How come cmake only adds `-mwasm64` to the link command when cross building? Is it possible to force it when not using conan's cross building support?
> 
I added a check around things not to call if emscripten is not set as os.

> 
> I am asking because it caches in the package folder.
> To give a counter example: I am recently toying a bit with writing Android apps. Its build system, gradle, caches libraries in `~/.gradle`.

I think it is fine to have caching in the package. I do not understand the gradle example for me it is another dependency manager which caches its dependencies in a folder in home. I do not see a difference with conan here and what this has to do about caching in package.

jgsogo commented on recipes/emsdk/all/conanfile.py:
```diff
@@ -0,0 +1,124 @@
+    homepage = "https://github.com/kripken/emscripten"
+    topics = ("conan", "emsdk", "emscripten", "sdk")
+    license = "MIT"
+    settings = "os"
```
Probably we need `arch` here too. We are now building Apple M1 as well (`MacOS` + `armv8`). I don't now if there are working binaries for it, but probably regular MacOS ones won't work, right?

werto87 replied:
I do not know about MacOs.

jgsogo commented on recipes/emsdk/all/conanfile.py:
```diff
@@ -0,0 +1,124 @@
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def package_id(self):
+        del self.info.settings.os
```
No need to modify anything here. We likely need different binaries (different package ID) for Windows, Linux, Macos,  M1,...

jgsogo reviewed (commented):
Letś try to move this forward, I would really like to start testing some libraries using this package as build-requires.

By the way, have your tried if the recipe also works for other emsdk versions? I'm interested in `1.39.8`, it is the one I've used previously to compile Qt

madebr commented:
@werto87 
Please consider https://github.com/werto87/conan-center-index/pull/1

madebr commented:
In the pr, I copied the `.pc` files to the package folder because of the error below.
Perhaps there is a workaround for this. Maybe creating a dummy pkgconfig folder?
```
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
shared:INFO: (Emscripten: Running sanity checks)
cache:INFO: generating system headers: sysroot_install.stamp... (this will be cached in "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/.emscripten_cache/sysroot_install.stamp" for subsequent builds)
Traceback (most recent call last):
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/em++.py", line 14, in <module>
    sys.exit(emcc.run(sys.argv))
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/emcc.py", line 1069, in run
    linker_inputs = phase_compile_inputs(options, state, newargs, input_files)
  File "/usr/lib64/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/emcc.py", line 2311, in phase_compile_inputs
    system_libs.ensure_sysroot()
  File "/usr/lib64/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/tools/system_libs.py", line 2009, in ensure_sysroot
    shared.Cache.get('sysroot_install.stamp', install_system_headers, what='system headers')
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/tools/cache.py", line 165, in get
    creator(cachename)
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/tools/system_libs.py", line 1992, in install_system_headers
    copytree_exist_ok(pkgconfig_src, pkgconfig_dest)
  File "/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/tools/system_libs.py", line 1960, in copytree_exist_ok
    for entry in os.scandir(src):
FileNotFoundError: [Errno 2] No such file or directory: '/home/maarten/.conan/data/emsdk/2.0.26/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/upstream/emscripten/system/lib/pkgconfig'
gmake[2]: *** [CMakeFiles/test_package.dir/build.make:76: CMakeFiles/test_package.dir/test_package.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/test_package.dir/all] Error 2
```

werto87 commented:
Hi,

I am only interested in providing the emsdk recipe for version 2.0.10 and I think this pullrequest contains this. Please consider merging this and do another pullrequest to enhance this recipe.

jgsogo commented:
> Hi,
> 
> I am only interested in providing the emsdk recipe for version 2.0.10 and I think this pullrequest contains this. Please consider merging this and do another pullrequest to enhance this recipe.

This is fine, @werto87 , we can add other versions afterward, specially if they do not work out-of-the-box.

Please, consider my review, as it was not related to other versions. Meanwhile, I'll trigger this PR once again, it's been one month since it last run, and something might change in the middle.

madebr commented:
@werto87 
Can you at least pull the test package from my PR?
Right now, the test package is not doing anything.

werto87 commented:
> > Hi,
> > I am only interested in providing the emsdk recipe for version 2.0.10 and I think this pullrequest contains this. Please consider merging this and do another pullrequest to enhance this recipe.
> 
> This is fine, @werto87 , we can add other versions afterward, specially if they do not work out-of-the-box.
> 
> Please, consider my review, as it was not related to other versions. Meanwhile, I'll trigger this PR once again, it's been one month since it last run, and something might have change during that time.



> @werto87
> Can you at least pull the test package from my PR?
> Right now, the test package is not doing anything.

After I fixed jgsogo's remarks and the build is green I will pull it.

madebr commented on recipes/emsdk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,24 @@
+class TestPackageConan(ConanFile):
+    settings = "os", "build_type", "arch", "compiler"
+    generators = "cmake"
+    test_type = "build_requires"
```
This is the reason why the M1 build is failing.
This is essentially requiring c3i to build emsdk twice: once for the build machine (=x86_64) and once for the host (=armv8).

I think, with the current behavior of conan/c3i of `test_type = "build_requires"`, it should not be used for the moment.

The only case where it is possible to be used if a package is os/arch/compiler-agnostic.

prince-chrismc commented on recipes/emsdk/all/conanfile.py:
```diff
@@ -0,0 +1,124 @@
+                self._chmod_plus_x("python_selector")
+            self._chmod_plus_x('emsdk')
+
+            self.run('%s install %s' % (emsdk, self.version))
```
Should this be in the package method?

werto87 commented:
Missing conan-center-bot comment. I cannot see why this commit cannot be built.

prince-chrismc commented:
Because it didnt run correctly.. no logs seem to be generated

You'll need to retrigger CI, close the pr wait 10s and then re-open it 🔁

jgsogo commented:
Sometimes it happens. I've triggered it manually.

prince-chrismc commented:
This downloads precompiled binaries https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#what-is-the-policy-on-creating-packages-from-pre-compiled-binaries

```
----Running------
> ./emsdk install 2.0.10
-----------------
Installing SDK 'sdk-releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'..
Installing tool 'node-14.15.5-64bit'..
Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/node-v14.15.5-darwin-x64.tar.gz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v14.15.5-darwin-x64.tar.gz, 31560456 Bytes
 [----------------------------------------------------------------------------]
Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/node-v14.15.5-darwin-x64.tar.gz' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/node/14.15.5_64bit'
Done installing tool 'node-14.15.5-64bit'.
Installing tool 'python-3.9.2-1-64bit'..
Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/python-3.9.2-1-macos-x86_64.tar.gz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/python-3.9.2-1-macos-x86_64.tar.gz, 32022000 Bytes
 [----------------------------------------------------------------------------]
Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/python-3.9.2-1-macos-x86_64.tar.gz' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/python/3.9.2-1_64bit'
Done installing tool 'python-3.9.2-1-64bit'.
Installing tool 'releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'..
Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/37fc7647c754ac9a28ad588c143b82286de0ef71-wasm-binaries.tbz2 from https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/37fc7647c754ac9a28ad588c143b82286de0ef71/wasm-binaries.tbz2, 265984173 Bytes
 [----------------------------------------------------------------------------]
Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/37fc7647c754ac9a28ad588c143b82286de0ef71-wasm-binaries.tbz2' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/upstream'
Done installing tool 'releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'.
Done installing SDK 'sdk-releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'.
```

prince-chrismc commented on recipes/emsdk/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,8 @@
+
+int main()
+{
+    std::cout << "Bincrafters\n";
+    return EXIT_SUCCESS;
```
Is this file used?

werto87 replied:
This is used to test the package. See conanfile in testpackage.

prince-chrismc reviewed (commented):
Why does the test package compile something?

werto87 commented:
> This downloads precompiled binaries https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#what-is-the-policy-on-creating-packages-from-pre-compiled-binaries
> 
> ```
> ----Running------
> > ./emsdk install 2.0.10
> -----------------
> Installing SDK 'sdk-releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'..
> Installing tool 'node-14.15.5-64bit'..
> Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/node-v14.15.5-darwin-x64.tar.gz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v14.15.5-darwin-x64.tar.gz, 31560456 Bytes
>  [----------------------------------------------------------------------------]
> Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/node-v14.15.5-darwin-x64.tar.gz' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/node/14.15.5_64bit'
> Done installing tool 'node-14.15.5-64bit'.
> Installing tool 'python-3.9.2-1-64bit'..
> Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/python-3.9.2-1-macos-x86_64.tar.gz from https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/python-3.9.2-1-macos-x86_64.tar.gz, 32022000 Bytes
>  [----------------------------------------------------------------------------]
> Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/python-3.9.2-1-macos-x86_64.tar.gz' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/python/3.9.2-1_64bit'
> Done installing tool 'python-3.9.2-1-64bit'.
> Installing tool 'releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'..
> Downloading: /Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/37fc7647c754ac9a28ad588c143b82286de0ef71-wasm-binaries.tbz2 from https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/37fc7647c754ac9a28ad588c143b82286de0ef71/wasm-binaries.tbz2, 265984173 Bytes
>  [----------------------------------------------------------------------------]
> Unpacking '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/zips/37fc7647c754ac9a28ad588c143b82286de0ef71-wasm-binaries.tbz2' to '/Users/jenkins/w/BuildSingleReference@8/.conan/data/emsdk/2.0.10/_/_/build/46f53f156846659bf39ad6675fa0ee8156e859fe/source_subfolder/upstream'
> Done installing tool 'releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'.
> Done installing SDK 'sdk-releases-upstream-37fc7647c754ac9a28ad588c143b82286de0ef71-64bit'.
> ```

It would be better to build from sources, but I lack motivation and I hope that this recipe can still be merged and building from sources can be added later.

jgsogo commented:
IMHO, there is a huge value in having this recipe available in ConanCenter. It will _always_ be used as a build-requires so having different versions for requirements should not be a huge problem. I totally agree and support that it's better to use packages provided by Conan, but some of them might take too long (`python`).

So I would vote for:
 * Keep downloading `python`, it's not a recipe that will be available in a couple of weeks, it will probably take more time. We can add it in the future.
 * Try to use `nodejs` from Conan: https://conan.io/center/nodejs. It is already available, so it should be fairly easy to try and see if it works, right? Maybe we need to add version `14.xx`, but recipe will probably work.

werto87 commented:
> IMHO, there is a huge value in having this recipe available in ConanCenter. It will _always_ be used as a build-requires so having different versions for requirements should not be a huge problem. I totally agree and support that it's better to use packages provided by Conan, but some of them might take too long (`python`).
> 
> So I would vote for:
> 
>     * Keep downloading `python`, it's not a recipe that will be available in a couple of weeks, it will probably take more time. We can add it in the future.
> 
>     * Try to use `nodejs` from Conan: https://conan.io/center/nodejs. It is already available, so it should be fairly easy to try and see if it works, right? Maybe we need to add version `14.xx`, but recipe will probably work.

I tried it when I started with the recipe I forgot exactly what  the problem was. I think it was something with calling 'emsdk install' and tricking the install script not to download the nodejs. The problem is that we would have to do this for every version. Maybe emsdk 2.0.10 downloads 14.xy and 2.0.11 14.zz. So it could be done for 2.0.10 but it needs a lot of work to work for every version and the emscripten team releases a lot of versions.

I think there should be a later version of this recipe where this is all handled. Maybe build from sources? . Its maybe not good right now but we can change it later and recipes which depending on this recipe do not have to change.

SSE4 reviewed (dismissed):
1. about `python` - I don't see huge value in downloading one more python interpreter. we (in many recipes, e.g. `meson`) assume python is already available, otherwise many recipes won't work. I know there is a pyinstaller, but it's more like a corner case. anyway, I don't think `emsdk` is one who should be responsible for that.

2. about `nodejs`. there is a way to skip download: https://github.com/bincrafters/conan-emsdk_installer/blob/testing/2.0.4/conanfile.py#L61. I don't like it, maybe there is a better way, but it may reduce download times and amount of nodejs installations on the machine.

perhaps, it's worth asking in upstream how to properly handle that, e.g. skip downloads of tools and specify to use other installations.

for the conan-center, I find it as a useful recipe, and already valuable enough to be included. we can do other improvements in follow-up PRs eventually.

jgsogo commented:
I've proposed a PR for this branch here https://github.com/werto87/conan-center-index/pull/2 using `nodejs` from Conan package (latest version).

Using those changes and some more changes here (https://github.com/conan-io/conan-center-index/compare/master...jgsogo:emsdk?expand=1) I'm able to use `corrade` and `magnum` with Emscripten, which is something that [@werto87 will like too](https://github.com/helmesjo/conan-magnum/issues/6) 😊

prince-chrismc commented on recipes/emsdk/config.yml:
```diff
@@ -0,0 +1,3 @@
+versions:
+  "2.0.26":
```
The number of times I've made this mistake... just an insurmountable number of failures  :)

```suggestion
  "2.0.30":
```

jgsogo commented on recipes/emsdk/all/conanfile.py:
```diff
@@ -0,0 +1,139 @@
+        with open(os.path.join(self.build_folder, self._source_subfolder, "emsdk_manifest.json"), 'r') as f:
+            data = json.load(f)
+            tools = data["tools"]
+            python = next((it for it in tools if (it["id"] == "python" and not it.get("is_old", False))), None)
+            ret["python"] = "python-{}-64bit".format(python["version"])
```
```suggestion
            if self.settings.os == "Linux":
                python = next((it for it in tools if (it["id"] == "python" and not it.get("is_old", False))), None)
                ret["python"] = "python-{}-64bit".format(python["version"])
```

Looks like Python is not installed by `emsdk` in Linux: https://emscripten.org/docs/getting_started/downloads.html#linux

jgsogo commented:
This is probably one of the recipes that will require exceptions in the hooks (https://github.com/conan-io/hooks). This is a compiler and for sure it is expected that it will package lots of these files. Feel free to open a PR to the hooks repo or we will take care of it in the following days (ping @uilianries ).

werto87 commented:
> This is probably one of the recipes that will require exceptions in the hooks (https://github.com/conan-io/hooks). This is a compiler and for sure it is expected that it will package lots of these files. Feel free to open a PR to the hooks repo or we will take care of it in the following days (ping @uilianries ).

Thank You! I will wait.

uilianries commented:
Exception case created: https://github.com/conan-io/hooks/pull/348

jgsogo commented:
@uilianries , this PR requires some more exceptions on the hooks side for `.pc` files

SSE4 commented on recipes/emsdk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,30 @@
+        return getattr(self, "settings_build", self.settings)
+
+    def build_requirements(self):
+        if self._settings_build.os == "Windows":
```
```suggestion
        self.build_requires("cmake/3.18.2")
        if self._settings_build.os == "Windows":
```

jgsogo replied:
Is 3.18 required? I think we should modify the limit in the `CMakeLists.txt`, not add the build-requires..

SSE4 replied:
if 3.18 is really required by this version of Emscripten, we have no choice except using build_requires

jgsogo commented on recipes/emsdk/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.18.2)
```
In ConanCenter we will converge to the minimum requirements for Conan v2, to maximize compatibility of recipes with the next Conan versions. Requirement for Conan v2 is CMake 3.15 (https://github.com/conan-io/tribe/blob/main/design/004-tools-cmake.md), so we try to ensure our recipes works with it.

Here I don't see anything requiring that CMake version, so it is probably safe to use something lower

SSE4 commented on recipes/emsdk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,31 @@
+        if self._settings_build.os == "Windows":
+            self.build_requires("make/4.3")
+
+    def build(self):
```
```suggestion
    def build(self):
        self.run("emcc -v", run_environment=True)
        self.run("em++ -v", run_environment=True)
```

SSE4 commented on recipes/emsdk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,33 @@
+        if self._settings_build.os == "Windows":
+            self.build_requires("make/4.3")
+
+    def build(self):
```
```suggestion
    def build(self):
        clang = os.path.join(self.deps_cpp_info["emsdk"].rootpath, "bin", "upstream", "bin", "clang")
        if self.settings.os == "Macos":
            self.run("file %s" % clang)
            self.run("otool -falh %s" % clang)
```

SSE4 commented on recipes/emsdk/all/conanfile.py:
```diff
@@ -0,0 +1,140 @@
+                raise ConanInvalidConfiguration("When using {}, target os must be Emscripten".format(self.name))
+
+    def package_id(self):
+        del self.info.settings.os
```
```suggestion
        pass
        #del self.info.settings.os
```

SSE4 replied:
it causes to run incorrect executable - ELF is consumed on Mac, it shouldn't happen

jgsogo commented:
I'm having a look at this PR. Trying to polish it for Macos world (running locally 💪 )

jgsogo commented:
One more hook... we've never reached that far (Windows builds) 🚀 

```
[HOOK - conan-center.py] post_package(): ERROR: [MS RUNTIME FILES (KB-H021)] The conan-center repository doesn't allow Microsoft Visual Studio runtime files. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H021) 
[HOOK - conan-center.py] post_package(): ERROR: [MS RUNTIME FILES (KB-H021)] Found files: ./bin/python/3.9.2-1_64bit/vcruntime140.dll; ./bin/python/3.9.2-1_64bit/vcruntime140_1.dll (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H021) 
ERROR: [HOOK - conan-center.py] post_package(): Some checks failed running the hook, check the output
```

jgsogo reviewed (approved):
🎉

werto87 commented:
> 
> 
> Thanks a lot, @werto87
> 
> During the weekend I kept working with `magnum` and I realized that it is using SDL1 for `emscripten` builds. As SDL is available in the `emsdk` it is not a big deal, but if we want to consume packages from ConanCenter for emscripten builds (#7427) we need to migrate `magnum` to use SDL2 or figure out how to work with `sdl12-compat` (a compatibility layer to provide SDL1 to SDL2 library, and help during the transition).
> 
> I don't know if you have any insights about this new topic, any idea if newer versions of `magnum` will use only SDL2? I have no knowledge at all of SDL so it is very hard for me to port the project to SDL2 😞

Thank you too,
to be honest I did not know (or forgot) about SDL1 and 2 before you mentiont it 😄 so I think I can not help. I think you should not worry about the next version of magnum because if something breaks in it we can fix it.
Please let me know when you managed to get magnum integration with imgui build for wasm working. I whould then test it with my project and help with bug fixes.

jgsogo commented:
@werto87 . The issue is that here I need to talk as a _package manager advocate_ (this is ConanCenter). I can build `magnum` using this emsdk recipe and it works almost out of the box... but only if I consume `sdl` from the emscripten port (it is already distributed with `emsdk`) because it provides all the patches required... but it is not SDL2 recipe and package provided by ConanCenter.
On the other hand, SDL2 builds using this emsdk recipe as well, but `magnum` is not prepared to use SDL2.

jgsogo commented:
You need to disable some components, you need to modify `magnum` recipe in this direction: https://github.com/conan-io/conan-center-index/pull/7412. As you can see, many components are disabled/deactivated for `emscripten` but at least `sdl2application` (using SDL2 embedded in `emsdk`) and `emscriptenapplication` are built.

It works, I promise, I did it live here: https://youtu.be/PjiJ-3sxKbs. Anyway I plan to write a blogpost with detailed information, and future work to be done.

jgsogo commented:
Hi! I've finally recreated my blog from scratch... and learned some NuxtJS in the process (still a lot of details to improve, I'm really bad at CSS) 🚀   Here you have the blogpost (most like a draft/quick notes): https://jgsogo.es/blog/2021-11-10-emscripten-magnum/

werto87 commented:
Hi, thanks for the credits. It was a good read. I will try it out.

werto87 commented:
I tried it on arch linux and it works perfectly :+1: .
Is there a reason why you show an image instead of the same image as wasm at the beginning?
